### ISSUE TITLE
Call to cuInit results in UNKNOWN_CUDA_ERROR

### ISSUE THREAD

**Issue by yxnal** [rapidsai/cudf#432]
I'm doing the following steps but having difficulty running tests about cudf.
# pull image
docker pull rapidsai/rapidsai:ubuntu1604_cuda92_py35
# run and activate
nvidia-docker  run -ti rapidsai/rapidsai:ubuntu1604_cuda92_py35 bash
cd rapids && source activate gdf
# install cudf
conda install -c nvidia -c rapidsai -c numba -c conda-forge -c defaults cudf=0.3.0
# run test
import numpy as np
from cudf.dataframe.dataframe import DataFrame
df = DataFrame()
df['keys'] = np.array([123], dtype=np.float64)
# test output
Traceback (most recent call last):
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/driver.py", line 227, in initialize
    self.cuInit(0)
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/driver.py", line 290, in safe_cuda_api_call
    self._check_error(fname, retcode)
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/driver.py", line 325, in _check_error
    raise CudaAPIError(retcode, msg)
numba.cuda.cudadrv.driver.CudaAPIError: [-1] Call to cuInit results in UNKNOWN_CUDA_ERROR

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/conda/envs/gdf/lib/python3.5/site-packages/cudf/dataframe/dataframe.py", line 236, in __setitem__
    self.add_column(name, col)
  File "/conda/envs/gdf/lib/python3.5/site-packages/cudf/dataframe/dataframe.py", line 546, in add_column
    series = self._prepare_series_for_add(data, forceindex=forceindex)
  File "/conda/envs/gdf/lib/python3.5/site-packages/cudf/dataframe/dataframe.py", line 519, in _prepare_series_for_add
    self._sanitize_columns(col)
  File "/conda/envs/gdf/lib/python3.5/site-packages/cudf/dataframe/dataframe.py", line 478, in _sanitize_columns
    series = Series(col)
  File "/conda/envs/gdf/lib/python3.5/site-packages/cudf/dataframe/series.py", line 74, in __init__
    data = columnops.as_column(data)
  File "/conda/envs/gdf/lib/python3.5/site-packages/cudf/dataframe/columnops.py", line 177, in as_column
    data = as_column(rmm.to_device(arbitrary))
  File "/conda/envs/gdf/lib/python3.5/site-packages/librmm_cffi/wrapper.py", line 228, in to_device
    to = self.device_array_like(ary, stream=stream)
  File "/conda/envs/gdf/lib/python3.5/site-packages/librmm_cffi/wrapper.py", line 218, in device_array_like
    stream=stream)
  File "/conda/envs/gdf/lib/python3.5/site-packages/librmm_cffi/wrapper.py", line 206, in device_array
    finalizer=self._make_finalizer(addr, stream))
  File "/conda/envs/gdf/lib/python3.5/site-packages/librmm_cffi/wrapper.py", line 158, in _array_helper
    ctx = cuda.current_context()
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/devices.py", line 194, in get_context
    return _runtime.get_or_create_context(devnum)
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/devices.py", line 162, in get_or_create_context
    return self.push_context(self.gpus[devnum])
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/devices.py", line 40, in __getitem__
    return self.lst[devnum]
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/devices.py", line 26, in __getattr__
    numdev = driver.get_device_count()
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/driver.py", line 336, in get_device_count
    self.cuDeviceGetCount(byref(count))
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/driver.py", line 270, in __getattr__
    self.initialize()
  File "/conda/envs/gdf/lib/python3.5/site-packages/numba/cuda/cudadrv/driver.py", line 230, in initialize
    raise CudaSupportError("Error at driver init: \n%s:" % e)
numba.cuda.cudadrv.error.CudaSupportError: Error at driver init:
[-1] Call to cuInit results in UNKNOWN_CUDA_ERROR:

# about cuda and nvidia-driver
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2018 NVIDIA Corporation
Built on Tue_Jun_12_23:07:04_CDT_2018
Cuda compilation tools, release 9.2, V9.2.148

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.72       Driver Version: 410.72     
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P4            Off  | 00000000:00:06.0 Off |                  N/A |
| N/A   27C    P8     6W /  75W |      0MiB /  7611MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

**Comment by kkraus14** [comment]
@yxnal It's possible you're installing the CUDA 10 version of libcudf and libcudf_cffi. Can you change your install command to `conda install -c nvidia -c rapidsai/label/cuda9.2 -c numba -c conda-forge -c defaults cudf=0.3.0` and try again?

**Comment by kkraus14** [comment]
cc @mt-jones since I believe he's seen this before and has a workaround.

**Comment by MurrayData** [comment]
I hit this issue with the same Docker container and, despite trying various combinations of NVIDIA driver and CUDA versions to no avail, I'm compiling from source, rather than using Docker atm.

**Comment by yxnal** [comment]
> @yxnal It's possible you're installing the CUDA 10 version of libcudf and libcudf_cffi. Can you change your install command to `conda install -c nvidia -c rapidsai/label/cuda9.2 -c numba -c conda-forge -c defaults cudf=0.3.0` and try again?

Still the same error

**Comment by harrism** [comment]
Did you install the NVIDIA Driver on your Ubuntu system? You need to install driver >= 410 for CUDA 10, or Driver >= 396 for CUDA 9.2.

**Comment by yxnal** [comment]
I have installed NVIDIA driver = 396.44

> > @yxnal It's possible you're installing the CUDA 10 version of libcudf and libcudf_cffi. Can you change your install command to `conda install -c nvidia -c rapidsai/label/cuda9.2 -c numba -c conda-forge -c defaults cudf=0.3.0` and try again?
> 
> Still the same error


> Did you install the NVIDIA Driver on your Ubuntu system? You need to install driver >= 410 for CUDA 10, or Driver >= 396 for CUDA 9.2.

I have installed NVIDIA-driver = 396.44 for  CUDA 9.2

**Comment by yxnal** [comment]
This is the __CUDA Information__ section:
CUDA Information
Error: CUDA device intialisation problem. Message:Error at driver init:
[-1] Call to cuInit results in UNKNOWN_CUDA_ERROR:
Error class: <class 'numba.cuda.cudadrv.error.CudaSupportError'>

**Comment by harrism** [comment]
Thanks @MurrayData. @yxnal I have had similar problems where I had mis-installed drivers, due to installing previously from the .deb installer and switching to the .run installer. I had to carefully remove all traces of the  driver using apt tools and reinstall.

**Comment by kkraus14** [comment]
@yxnal are you still experiencing the issue?

**Comment by harrism** [comment]
@yxnal please respond if you still have this issue. Otherwise we will close this issue.

**Comment by harrism** [comment]
Closing. Please reopen if still an issue.

**Comment by paulcheck** [comment]
I am having the same problem on a Debian unstable machine running tensorflow-gpu via anaconda.  I have the Nvidia driver installed and nvidia-smi is reporting the driver is v 418 and cuda 10 capable.  When I run numba -s it gives the same error, even though I have cuda 10 installed in my conda environment.  I saw someone make a comment somewhere that this problem can be caused by the Nvidia drivers not located in a certain path, but I don't know what directory to add to the path variable.

**Comment by harrism** [comment]
We can't really support tensorflow here. Did you install the CUDA toolkit or only the conda package? I think you have to have both.

**Comment by paulcheck** [comment]
I'm not sure what you mean.  Within the conda environment, I have all necessary packages installed.  For example:

cudatoolkit               10.0.130                      0    anaconda
cudnn                     7.6.0                cuda10.0_0    anaconda

I also have cuda installed on my debian machine (I'm not sure if you're asking about packages that are installed outside of the conda environment, or not).  I'm under the impression that conda is a stand alone environment.

**Comment by harrism** [comment]
I'm not sure the conda cudatoolkit package actually installs all of the CUDA toolkit, or whether it depends on it it being installed. @kkraus14 would know better. I usually install the CUDA toolkit on my machine first, and then set up the conda environment. But I also don't know whether Tensorflow needs the CUDA toolkit.

**Comment by paulcheck** [comment]
When I install tensorflow-gpu in the conda environment, it automatically downloads a number of conda packages, including those two I referenced above.

Yes, I do have the cuda toolkit installed outside of conda, as a debian package.

**Comment by harrism** [comment]
OK, then I suggest reinstalling display drivers from scratch like @MurrayData mentioned worked for him.

**Comment by paulcheck** [comment]
Indeed, this worked.  Yes, the conda nvidia toolkit does depend on the system cuda library/toolkit.  That was the piece I was unaware of.  The reason for the error initially was that there was a version mismatch between those two, since Debian unstable system cuda was only at version 9.2, and of course conda was using 10.1.    In order to get this working, I had to dip into the experimental branch of the Debian repositories to get access to the 10.1 version of the cuda library/toolkit. Even doing that didn't work right away, since not all needed nvidia packages were being upgraded to version 10.1.  So, I removed all packages that had anything to do with nvidia/cuda, and then installed from the experimental branch.  At that point, numba -s was no longer reporting the error above.  My sample multi-gpu tensorflow program that was working on my Windows 10 conda installation now works on the Debian unstable/experimental conda installation.

Thx for the assist.  I wrote this detailed message to help others with a similar issue on a Debian machine.  If I were starting from scratch I would consider using Ubuntu 18.04 (version from memory), since it has cuda version 10 packages.  But, hopefully Debian unstable will catch up soon ;).
